retry/retry/src/lib.rs

```diff
@@ -0,0 +1,241 @@
+}
+
+/// Makes and, in case of failure, retries an idempotent call until instructed otherwise
+///
+/// This function is suitable for calls that can be safely retried without side effects.
+/// It will retry the call until either:
+/// - The call succeeds
+/// - The retry condition returns false
+/// - A non-retryable error occurs
+///
+/// # Arguments
+///
+/// * `call` - The (idempotent) call to execute and retry if needed
+/// * `stop_trying` - A function that determines when to stop (re)trying the call
+///
+/// # Returns
+///
+/// * `Ok(Response)` if the call succeeds
+/// * `Err(RetryError)` if the call fails and cannot be retried
+pub async fn call_idempotent_method_with_retry<'a, 'm, P>(
+    call: Call<'a, 'm>,
+    stop_trying: &mut P,
+) -> Result<Response, RetryError>
+where
+    P: FnMut() -> bool,
+{
+    let mut no_unknown_results = true;
+
+    loop {
+        if stop_trying() {
+            return Err(if no_unknown_results {
+                RetryError::CallFailed(ErrorCause::GaveUpRetrying)
+            } else {
+                RetryError::StatusUnknown(ErrorCause::GaveUpRetrying)
+            });
+        }
+
+        match call.clone().await {
+            Ok(result) => return Ok(result),
+            Err(e) if !e.is_immediately_retryable() => {
+                if no_unknown_results {
+                    return Err(RetryError::CallFailed(ErrorCause::CallFailed(e)));
+                } else {
+                    return Err(RetryError::StatusUnknown(ErrorCause::CallFailed(e)));
+                }
+            }
+            Err(e) if !e.is_clean_reject() => {
+                no_unknown_results = false;
+                continue;
+            }
+            // The only remaining option is a non-sync SysTransient => retry
+            Err(_e) => continue,
+        }
+    }
+}
+
+/// Makes and, in case of failure, retries a non-idempotent call until instructed otherwise
+///
+/// This function is suitable for calls that may have side effects and should be
+/// retried with caution. It will retry the call until either:
+/// - The call succeeds
+/// - The retry condition returns false
+/// - A non-retryable error occurs
+/// - An error occurs where we cannot determine the final status of the call
+///
+/// # Arguments
+///
+/// * `call` - The call to retry
+/// * `stop_trying` - A function that determines whether to stop (re)trying the call
+///
+/// # Returns
+///
+/// * `Ok(Response)` if the call succeeds
+/// * `Err(RetryError)` if the call fails and cannot be retried
+pub async fn call_nonidempotent_method_with_retry<'m, 'a, P>(
+    call: Call<'m, 'a>,
+    stop_trying: &mut P,
+) -> Result<Response, RetryError>
+where
+    P: FnMut() -> bool,
+{
+    loop {
+        if stop_trying() {
+            return Err(RetryError::CallFailed(ErrorCause::GaveUpRetrying));
+        }
+
+        match call.clone().await {
+            Ok(res) => return Ok(res),
+            Err(e) if !e.is_immediately_retryable() => {
```

Comment:
even if the error cannot be retried immediately, we could still make a dummy inter-canister call (e.g., calling the canister_status endpoint of the mgmt canister) to retry in a new message; of course, this is a busy waiting but without hitting the instruction limit

Reply:
If it's a synchronous error, we could still hit the instruction limit if we retried in a loop. But even for asynchronous errors, the "not immediately retriable errors" really don't have a good chance of succeeding upon immediate (or near-immediate) retry; e.g., `CanisterError` or `CanisterReject` aren't in general likely to be fixed if we retry within a few seconds. Maybe if we ever implement the `sleep` API and finer grained error code we could reconsider, e.g., a canister out of cycles or canister stopping might be sensible retried with an exponential backoff policy.

Reply:
> If it's a synchronous error, we could still hit the instruction limit if we retried in a loop.

not if the loop has the form
```
loop {
  // do something
  ic_cdk::call(aaaaa-aa, canister_status, ...).await;
}
```

Reply:
I'd expect that synchronous error due to a queue being full might have a good chance of succeeding if retried immediately (in a new message), e.g., because the target canister was executed and picked something from the queue (this is why the asynchronous _transient_ errors are retried, expecting that the remote subnet is not low on resources after a little while).

Of course, the dummy call might have a side-effect (filling up a different queue) so I'm not saying this should be done in the very same way I sketched it above.

Reply:
But a synchronous error might also happen because the subnet's out of message memory, you hit the callback limit, etc. Retrying in a loop is a pretty bad idea in that case. One could bail out if the canister status call fails, but even if it succeeds, if parts of the system are under heavy load, it's likely not great to add to that load quickly. The async errors would at least likely cause us to back off for a while (since they'd hopefully take some time to trigger), though, again, a `sleep` API would be a much more sensible solution.